ts/calculator.ts

```diff
@@ -0,0 +1,476 @@
+            symbol = "ln(";
+            formula = button.formula + "(";
+
+            data.operation.push(symbol);
+            data.formula.push(formula);
+        } else if (button.name === "log") {
+            symbol = "log(";
+            formula = button.formula + "(";
+
+            data.operation.push(symbol);
+            data.formula.push(formula);
+        } else if (button.name === "exp") {
+            symbol = "exp(";
+            formula = button.formula + "(";
+
+            data.operation.push(symbol);
+            data.formula.push(formula);
+        } else {
+            symbol = button.symbol + "(";
+            formula = button.formula;
+            data.operation.push(symbol);
+            data.formula.push(formula);
+        }
+    } else if (button.type === "key") {
+        if (button.name === "clear") {
+            data.operation = [];
+            data.formula = [];
+            updateOutputresult('0');
+        } else if (button.name === "delete") {
+            data.operation.pop();
+            data.formula.pop();
+        } else if (button.name === "deg") {
+            RADIAN = true;
+            angleToggle();
+
+        } else if (button.name === "rad") {
+            RADIAN = false;
+            angleToggle();
+        }
+
+    } else if (button.type === "calculate") {
+        let formula_str = data.formula.join("");
+        let POWER_SEARCH_RESULT = search(data.formula, POWER);
+        let FACTORIAL_SEARCH_RESULT = search(data.formula, FACTORIAL);
+        const BASES = powerbasegetter(data.formula, POWER_SEARCH_RESULT);
+        const NUMBERS = factorialnumgetter(data.formula, FACTORIAL_SEARCH_RESULT);
+
+        BASES.forEach((base) => {
+            let toreplace = base + POWER;
+            let replacement = "Math.pow(" + base + ",";
+
+            formula_str = formula_str.replace(toreplace, replacement);
+        });
+
+        NUMBERS.forEach((number) => {
+            formula_str = formula_str.replace(number.toReplace, number.replacement);
+        });
+
+        try {
+            result = Number(eval(formula_str));
+            var decCount = decimalCount(result);
+            if (decCount > 5) {
+                result = (parseFloat(result.toPrecision(5)));
+            }
+        } catch (error) {
+            if (error instanceof SyntaxError) {
+                result = "SyntaxError";
+                updateOutputresult(result);
+                return;
+            }
+        }
+
+        ans = result;
+        data.operation = [result];
+        data.formula = [result];
+        updateOutputresult(result);
+        return;
+    }
+    updateOutputOperation(data.operation.join(""));
+}
+interface ReplaceType {
+    toReplace: string,
+    replacement: string
+}
+function factorialnumgetter(formula: (string | number | boolean)[], FACTORIAL_SEARCH_RESULT: number[]) {
+    let numbers: ReplaceType[] = [];
+
+    FACTORIAL_SEARCH_RESULT.forEach((fact_index) => {
+        let number = [];
```

Comment:
Add type definition.
number array is inferred as `any[]`.
Also, number is a reserved keyword in typescript.